### ISSUE TITLE
Multi language support

### ISSUE THREAD

**Issue by Amost-** [OpenEmu/OpenEmu#1083]
Some games read the default system language and the game's one automatically [Nintendo DS].
In some games you can't change the language manually.
This feature would be much appreciated. The core you use for NDS, desmume, has this support.

**Comment by dunkler-schatten** [comment]
I'm from Germany
And Zelda on DS in English is not playable for me. :-(
Please make a option to change the Language in the Nintendo DS Core. Please!!!!! And other Cores. Big Thanks!!

**Comment by maki-chan** [comment]
I had this problem myself and wasn't able to compile the DeSmuME core for OpenEmu myself, so I decided to dig deeper and I disassembled the executable that's used in the Core. I then looked up in the official DeSmuME source code how the language flag is loaded and which language flag is German (turns out it's 3).

Then I fired up my hex editor and changed the instruction to directly store 3 into the firmware language variable instead of the `fw_config_data` value (which defaults to 1 for English).

TL;DR: If you want to have German language for your DS games, you can replace your DeSmuME core in `~/Library/Application Support/Cores/DeSmuME.oecoreplugin` with the one you can find [in this zip file](https://app.box.com/v/DeSmuME-German-Core).

**Comment by dunkler-schatten** [comment]
maki-chan 1000 Thanks!! its work perfectly

**Comment by clobber** [comment]
@maki-chan Thanks, but I removed your link to the core file download. You can make guides on how to do that, but for many reasons I would prefer we don't share builds on here that weren't compiled by OE devs.

**Comment by maki-chan** [comment]
In that case, everybody could do the following (a guide is even better, as I can tell everybody to choose a language they want):

1. Open the file `~/Library/Application Support/OpenEmu/Cores/DeSmuME.oecoreplugin/Contents/MacOS/DeSmuME` in a hex editor, e.g. Hex Fiend
2. Search for offset `693512` or `0xA9508`, the following 3 bytes should be `8A 43 53` and those are the bytes you have to replace with `B0 ?? 90`<sup>1) 2)</sup>
3. Save the new executable with the changed bytes

<sup>1)</sup>Replace the `??` with the following bytes to enable a specific language:
- `00` => Japanese
- `01` => English
- `02` => French
- `03` => German
- `04` => Italian
- `05` => Spanish

<sup>2)</sup> A more technical explanation of what you are doing here if you are interested in it. Usually, this command means that it moves a specific byte at a memory address (`rbx+0x53` in this case, whereas `43`stands for the `rbx` register and `53` for the `0x53` offset; `rbx` holds the `fw_config_data` variable address) to the accumulator low byte register (`8A` represents that, where `8x` is the move command and `A` says "move to `al`"). Once it is changed, it says "move an immediate byte value to `al`" (that's the `B0` byte) and then we give it the byte value that should be stored (which is exactly what the `--fwlang=?`value would be if we'd start DeSmuME via CLI). The `90` byte at the end stands for "no operation" and is needed so we don't change the file size and offsets, etc. of every other code that comes after our changed instruction.

**Comment by Marcounet** [comment]
Hi!
I tried this extreme solution, but Hex Fiend didn't found "693512" or 30xA9508" offset. (T_T)

**Comment by maki-chan** [comment]
Hi Marcounet,

did you use the `Edit -> Jump to Offset` function for that? It should definitely find the offset in the DeSmuME file.

**Comment by Marcounet** [comment]
@maki-chan OK it worked. My mistake was using the "Find" function and not the "Jump" one. Thanks a lot and bests regards.

**Comment by ssaring** [comment]
@maki-chan  Thanks, works perfectly. Would be great when the next OpenEmu release could have a language selection in the preferences or could map the users system locale and pass it to the emulator engine.

**Comment by MauriceLetsPlay** [comment]
Can anyone code me make an SNES German core?
Sorry im Luxembourg, an i can't good english.

**Comment by 7h0b5** [comment]
Can someone do the same guide but for PPSSPP core ?
Thanks for the help tho.

**Comment by clockstroke** [comment]
DeSmuME seems to have changed due to updates. Offset 693512 is not the same anymore. Does anyone have an idea ?

**Comment by Tvaing** [comment]
As clockstroke, I didn't succeed in changing the language with this method because offset 693512 has changed.
With Plist Editor, I found a language selector in the file : _~/Library/Application Support/OpenEmu/Cores/DeSmuME.oecoreplugin/Contents/info.plist_ but it doesn't seems to change anything in-game.
Can anyone help us ? How could we find the good offset in the hexadecimal code ?

**Comment by Sillv** [comment]
@clockstroke and @Tvaing 
There is only one combo 8A 43 53 in the whole code, no matter what version you have.
That means you can easily Cmd + F on 8A 43 53 to find these bytes and replace it by B0 ?? 90.

It worked that way for me.

**Comment by Tvaing** [comment]
@Sillv It totally worked, a huge thank you !

**Comment by hazuna25** [comment]
@Sillv When I change the combo 8A 43 53 to B0 03 90 the emulator can't load the core anymore. I tried so many times. Is there another way to change the language into german? I'm tired of trying and failing...

**Comment by mansemino** [comment]
One way to select custom language for NDS games is by using cheat codes.
Just create a new cheat code and use the following code according to your language:

- English: 22FFFCE4 00000001
- French: 22FFFCE4 00000002
- German: 22FFFCE4 00000003
- Italian: 22FFFCE4 00000004
- Spanish: 22FFFCE4 00000005

Tested for spanish language in european/multi language games.
Hope it helps.

**Comment by macko777** [comment]
@EdgardPinget & @maki-chan 

Hi, I'm french. 

You make my day ;) i buy a ps3 controller 30€ in second hand (yeah that's too expensive but work perfectly...) and it doesn't work on desume but work on Open Emu but don't have french language... Now i have it, so Big Thanks to you for your helps !

**Comment by MoutonBinoclard** [comment]
Someone can say me where I must go for change the Desmume language ? I don't find it ...

**Comment by MoutonBinoclard** [comment]
I have tap 8A 43 53 for search a line in the code of desmume but I have find nothing ... Help ?

**Comment by Timeonfr** [comment]
Hi,

I'm on mac and I have OpenEmu. I'm desperately trying to get a DS game in EUR version to work in French but I can't do it. 

I tried your methods but at the moment of opening the file with Hex Fiend it doesn't work: "The document "DeSmuME.oecoreplugin" could not be opened. The file is a directory which is not a supported type. »

Would you have a solution for me?

**Comment by Timeonfr** [comment]
> Une façon de sélectionner une langue personnalisée pour les jeux NDS consiste à utiliser des codes de triche.
> Créez simplement un nouveau code de triche et utilisez le code suivant en fonction de votre langue:
> 
> * Anglais: 22FFFCE4 00000001
> * Français: 22FFFCE4 00000002
> * Allemand: 22FFFCE4 00000003
> * Italien: 22FFFCE4 00000004
> * Espagnol: 22FFFCE4 00000005
> 
> Testé pour la langue espagnole dans des jeux européens / multilingues.
> J'espère que cela aide.

I can't do it, even in the Spanish version. When I put "22FFFCE4 00000002" for example, and I click on "Activate now" my game does not change to the .FR version :(

**Comment by Timeonfr** [comment]
I don't know if this will help some of you but I'm posting how I did it on this 25th February 2021!

I went as previously proposed here : ~/Library/Application Support/OpenEmu/Cores/DeSmuME.oecoreplugin

The DeSmuME.oecoreplugin file did not propose me to go further in Contents/MacOS/DeSmuME. When I wanted to open it with Hex Fiend, it gave me an error message.

I then changed the extension by clicking on Information and replacing .oecoreplugin by .gid.

Then I was able to go to Contents/MacOS/DeSmuME, and opened DeSmuME with Hex Fiend as advised.

I did CMD + F and I manually typed 8A 43 53 (copy/paste didn't help me find it), and I replaced it with B0 ?? 90 (for me it was B0 01 90 since I wanted the French version). 

Once this was done I saved, and put the extension .oecoreplugin back in place of .gid on the DeSmuME.gid file. I removed my game EUR version and reinstalled it, it worked.

Finally I just followed the tutorial above, only I couldn't understand why DeSmuME.oecoreplugin didn't work, because I was confusing DeSmuME in /Cores and DeSmuME in /MacOS. 

Many thanks! If some people are still having trouble, you can send me a message on Discord: Timeon#7987 (I don't accept friend requests but I get the messages).

**Comment by Dadoudoux** [comment]
@Timeonfr 
<img width="1440" alt="58C56354-DBF2-49FC-8F18-A7CC88099A33" src="https://user-images.githubusercontent.com/110688279/183125946-4a27ee95-f32a-4aea-b032-d6b944567c11.png">
Mon mac me dis ça quand je veux sauvegarder que dois-je faire

**Comment by LoKy-dev** [comment]
> One way to select custom language for NDS games is by using cheat codes. Just create a new cheat code and use the following code according to your language:
> 
> * English: 22FFFCE4 00000001
> * French: 22FFFCE4 00000002
> * German: 22FFFCE4 00000003
> * Italian: 22FFFCE4 00000004
> * Spanish: 22FFFCE4 00000005
> 
> Tested for spanish language in european/multi language games. Hope it helps.

@mansemino Thank you so much, i'm not confident enough to change bytes of executables without worrying about breaking everything and the offset specified in other comments it's not valid anymore due to DesmuMe changes.

This worked flawlessly for me (after restarting the game with the cheat active) and is so tiny yet elegant and efficient solution

**Comment by fhughjanus-sketch** [comment]
Hello, I've never used this before so I don't know how github or desmume works, or if this thing is still up for people to answer. Anyone there? Hello?
